contracts/v0.2/src/Feed.sol

```diff
@@ -0,0 +1,137 @@
+
+    struct Round {
+        int256 answer;
+        uint64 updatedAt;
+    }
+
+    error OracleAlreadyEnabled();
+    error OracleNotEnabled();
+    error NoDataPresent();
+
+    event OraclePermissionsUpdated(address indexed oracle, bool indexed whitelisted);
+    event FeedUpdated(int256 indexed answer, uint256 indexed roundId, uint256 updatedAt);
+
+    constructor(uint8 _decimals, string memory _description) Ownable(msg.sender) {
+        decimals = _decimals;
+        description = _description;
+    }
+
+    // TODO verification
+    function submit(int256 _answer) external {
+        uint64 roundId_ = latestRoundId + 1;
+
+	rounds[roundId_].answer = _answer;
+        rounds[roundId_].updatedAt = uint64(block.timestamp);
+
+	emit FeedUpdated(_answer, roundId_, block.timestamp);
+	latestRoundId = roundId_;
+    }
+
+    function changeOracles(
+        address[] calldata _removed,
+        address[] calldata _added
+    ) external onlyOwner {
+        for (uint256 i = 0; i < _removed.length; i++) {
+            removeOracle(_removed[i]);
+        }
+
+        for (uint256 i = 0; i < _added.length; i++) {
+	    if (_added[i] == address(0)) {
+		continue;
+	    }
+            addOracle(_added[i]);
+        }
+    }
+
+    function getOracles() external view returns (address[] memory) {
+        return oracles;
+    }
+
+    function latestRoundData()
+        external
+        view
+        virtual
+        override
+        returns (uint64 roundId, int256 answer, uint256 updatedAt)
+    {
+        return getRoundData(latestRoundId);
+    }
+
+    function latestRoundUpdatedAt() external view returns (uint256) {
+        Round storage round = rounds[latestRoundId];
+        return round.updatedAt;
+    }
+
+    function getRoundData(uint64 _roundId)
+        public
+        view
+        virtual
+        override
+        returns (uint64 roundId, int256 answer, uint256 updatedAt)
+    {
+        Round memory r = rounds[_roundId];
+
+        if (r.updatedAt == 0 || !validRoundId(_roundId)) {
+            revert NoDataPresent();
+        }
+
+        return (_roundId, r.answer, r.updatedAt);
+    }
+
+    function typeAndVersion() external pure virtual override returns (string memory) {
+        return "Feed v0.2";
+    }
+
+    function addOracle(address _oracle) private {
+        if (oracleEnabled(_oracle) != NOT_FOUND) {
+            revert OracleAlreadyEnabled();
+        }
+
+        oracles.push(_oracle);
+        emit OraclePermissionsUpdated(_oracle, true);
+    }
+
+    function removeOracle(address _oracle) private {
+	int256 oracleId = oracleEnabled(_oracle);
+        if (oracleId == NOT_FOUND) {
+            revert OracleNotEnabled();
+        }
+
+        address tail = oracles[oracles.length - 1];
+        oracles[uint256(oracleId)] = tail;
+	oracles.pop();
+
+        emit OraclePermissionsUpdated(_oracle, false);
+    }
+
+    function oracleEnabled(address _oracle) private view returns (int256) {
```

Comment:
It is because when `oracle` address is not found among registered ones, we have to represent it somehow. In this case, I chose to use -1. We could theoretically use the largest value (2^256 -1) to represent `not found` state, or return `index` and boolean flag whether the oracle was found.

Reply:
For such case, I used to set [0] index as `not found` value, and kept [0] index empty, made entries start from 1.

Reply:
If we go with that, it distorts both the use of the array and the returned index. Everywhere in the code, we have to keep it in mind. If you think that the current solution is not good enough, I think that having 2^256 -1  for `not found` would be a cleaner.